skimage/morphology/_misc_cy.pyx

```diff
@@ -0,0 +1,122 @@
+#cython: cdivision=True
+#cython: boundscheck=False
+#cython: nonecheck=False
+#cython: wraparound=False
+
+"""Cython code used in `remove_near_objects` function."""
+
+cimport numpy as cnp
+
+from .._shared.fused_numerics cimport np_anyint
+
+
+def _remove_near_objects(
+    np_anyint[::1] out not None,
+    Py_ssize_t[::1] border_indices not None,
+    Py_ssize_t[::1] inner_indices not None,
+    kdtree,
+    cnp.float64_t p_norm,
+    cnp.float64_t min_distance,
+    tuple shape,
+):
+    """Remove objects until a minimal distance is ensured.
+
+    Iterates over all objects (pixels that aren't zero) inside an image and
+    removes "nearby" objects until all remaining ones are spaced more than a
+    given minimal distance from each other.
+
+    Parameters
+    ----------
+    out :
+        An array with labels for each object in `image` matching it in shape.
+    border_indices, inner_indices :
+        Indices into `out` for the border of objects (`border_indices`) and
+        the inner part of objects (`inner_indices`). `border_indices`
+        determines the iteration order; objects that are indexed first are
+        preserved. Indices must be sorted such, that indices pointing to the
+        same object are next to each other.
+    kdtree : scipy.spatial.cKDTree
+        A KDTree containing the coordinates of all objects in `image`.
+    min_distance :
+        The minimal allowed distance between objects.
+    p_norm :
+        The Minkowski p-norm used to calculate the distance between objects.
+        Defaults to 2 which corresponds to the Euclidean distance.
+    shape :
+        The shape of the unraveled `image`.
+    """
+    cdef:
+        Py_ssize_t i_indices, j_indices  # Loop variables to index `indices`
+        Py_ssize_t i_out  # Loop variable to index `out`
+        np_anyint object_id, other_id
+        list neighborhood
+        set remembered_ids
+
+    remembered_ids = set()
+    for i_indices in range(border_indices.shape[0]):
+        i_out = border_indices[i_indices]
+        object_id = out[i_out]
+        # Skip if sample is part of a removed object
+        if object_id == 0:
+            continue
+
+        neighborhood = kdtree.query_ball_point(
+            kdtree.data[i_indices, ...],
```

Comment:
Just cross-reviewing from SciPy since there were some concerns expressed about the `list` return type here (which we need because of the ragged data structures that are possible) for performance reasons.

A few questions:

1) Do you have any sample timings/examples demonstrating the bottleneck/% time spent on this line for a typical use case?
2) Any chance you could use the `workers` argument to speed up via concurrency? This might be particularly viable if you could accumulate the query points externally first.
3) How many points are you querying against in a typical use case? The reason I ask is that you seem to be querying the tree against itself, which may benefit from using `query_ball_tree` instead.

Reply:
For the following case

```python
import numpy as np
import skimage as ski

image = ski.color.rgb2gray(ski.data.hubble_deep_field())
foreground1 = image > ski.filters.threshold_li(image)  # many small objects
foreground2 = np.ones_like(foreground1)
foreground2[400, :] = 0  # two very large objects
foreground = np.concatenate([foreground1, foreground2])
objects = ski.measure.label(foreground)

%timeit ski.morphology.remove_near_objects(objects, min_distance=5)
%timeit ski.morphology.remove_near_objects(objects, min_distance=50)
%timeit ski.morphology.remove_near_objects(objects, min_distance=100)
%timeit ski.morphology.remove_near_objects(objects, min_distance=300)
```

Querying and entire object's indices at once with `kdtree.query_ball_point(kdtree.data[start:stop], ...]`,

<details><summary>Patch</summary>

```patch
Subject: [PATCH] Query indices with the same object ID at once
---
Index: skimage/morphology/_misc_cy.pyx
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/skimage/morphology/_misc_cy.pyx b/skimage/morphology/_misc_cy.pyx
--- a/skimage/morphology/_misc_cy.pyx	(revision c8756cae67078c1cfb05763a503792e05e842aef)
+++ b/skimage/morphology/_misc_cy.pyx	(date 1713872591799)
@@ -46,34 +46,45 @@
         The shape of the unraveled `image`.
     """
     cdef:
-        Py_ssize_t i_indices, j_indices  # Loop variables to index `indices`
+        Py_ssize_t i_indices, j_indices, start, stop  # Loop variables to index `indices`
         Py_ssize_t i_out  # Loop variable to index `out`
-        np_anyint object_id, other_id
-        list neighborhood
+        np_anyint start_id, stop_id
         set remembered_ids
+        list n
 
     remembered_ids = set()
-    for i_indices in range(border_indices.shape[0]):
-        i_out = border_indices[i_indices]
-        object_id = out[i_out]
-        # Skip if sample is part of a removed object
-        if object_id == 0:
+    start = 0
+    start_id = out[border_indices[start]]
+    for stop in range(border_indices.shape[0]):
+        i_out = border_indices[stop]
+        stop_id = out[i_out]
+
+        if start_id == stop_id:
+            continue
+        elif start_id == 0:
+            start = stop
+            start_id = stop_id
             continue
 
         neighborhood = kdtree.query_ball_point(
-            kdtree.data[i_indices, ...],
+            kdtree.data[start:stop, ...],
             r=min_distance,
             p=p_norm,
         )
-        for j_indices in neighborhood:
-            # Check object IDs in neighborhood
-            other_id = out[border_indices[j_indices]]
-            if other_id != 0 and other_id != object_id:
-                # If neighbor ID wasn't already removed or is the current one
-                # remove the boundary and remember the ID
-                _remove_object(out, border_indices, j_indices)
-                remembered_ids.add(other_id)
+
+        for n in neighborhood:
+            for j_indices in n:
+                # Check object IDs in neighborhood
+                other_id = out[border_indices[j_indices]]
+                if other_id != 0 and other_id != start_id:
+                    # If neighbor ID wasn't already removed or is the current one
+                    # remove the boundary and remember the ID
+                    _remove_object(out, border_indices, j_indices)
+                    remembered_ids.add(other_id)
 
+        start = stop
+        start_id = out[border_indices[start]]
+
     # Delete inner parts of remembered objects
     for j_indices in range(inner_indices.shape[0]):
         object_id = out[inner_indices[j_indices]]

```

</details> 
 
is faster for `min_distance=5` and `50` but a lot slower once `min_distance` get's larger. Without profiling I'm not really sure about the reason. The implementation that queries single points only, still seems to be the one that's fastest overall.

Using `workers` seems to introduce an overhead that, for most cases, is slower regardless of the specific implementation.